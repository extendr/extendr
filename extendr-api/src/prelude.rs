//! # Common exports for extendr-api.
//!
//! This allows us to be more selective about exports and avoid users
//! using deprecated features.

pub use super::{
    new_owned, print_r_error, print_r_output, FromRobj, IsNA, RType, FALSE, NA_INTEGER, NA_LOGICAL,
    NA_REAL, NA_STRING, NULL, TRUE,
};

pub use super::error::{Error, Result};

pub use super::functions::{
    base_env, base_namespace, base_symbol, blank_scalar_string, blank_string, brace_symbol,
    bracket_2_symbol, bracket_symbol, class_symbol, current_env, device_symbol, dim_symbol,
    dimnames_symbol, dollar_symbol, dot_defined, dot_method, dot_package_name, dot_target,
    dots_symbol, double_colon_symbol, empty_env, eval_string, find_namespace, global_env,
    global_function, global_var, lastvalue_symbol, levels_symbol, local_var, missing_arg,
    mode_symbol, na_rm_symbol, na_str, na_string, name_symbol, names_symbol, namespace_env_symbol,
    namespace_registry, new_env, new_env_with_capacity, nil_value, package_symbol, parse,
    previous_symbol, quote_symbol, row_names_symbol, seeds_symbol, sort_list_symbol, source_symbol,
    spec_symbol, srcref, triple_colon_symbol, tsp_symbol, unbound_value,
};

pub use crate::{append, append_lang, append_with_name, args, call, lang, make_lang};
pub use crate::{
    data_frame, factor, global, list, r, reprint, reprintln, rprint, rprintln, sym, test, var, R,
};

pub use super::logical::Bool;

pub use super::matrix::{RArray, RColumn, RMatrix, RMatrix3D};

pub use super::robj::{IntoRobj, Robj, RobjItertools};

pub use super::thread_safety::{
    catch_r_error, handle_panic, single_threaded, this_thread_id, throw_r_error,
};

pub use super::wrapper::{
    Character, Env, Expr, Func, Lang, List, Nullable, Pairlist, Primitive, Promise, Raw, Symbol,
};

#[cfg(feature = "ndarray")]
pub use super::robj_ndarray::*;

#[cfg(feature = "ndarray")]
pub use ndarray::*;

pub use extendr_macros::{extendr, extendr_module};

pub use super::iter::{
    EnvIter, Int, ListIter, Logical, PairlistIter, PairlistTagIter, Real, StrIter,
};
