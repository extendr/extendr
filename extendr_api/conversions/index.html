<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `conversions` mod in crate `extendr_api`."><title>extendr_api::conversions - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="extendr_api" data-themes="" data-resource-suffix="" data-rustdoc-version="1.92.0-nightly (2aaa62b89 2025-10-24)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../../static.files/storage-e2aeef58.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">Module conversions</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../extendr_api/index.html"><img src="https://raw.githubusercontent.com/extendr/extendr/master/extendr-logo-256.png" alt="logo"></a><h2><a href="../../extendr_api/index.html">extendr_<wbr>api</a><span class="version">0.8.0</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module conversions</a></h2><h3><a href="#modules">Module Items</a></h3><ul class="block"><li><a href="#modules" title="Modules">Modules</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate extendr_<wbr>api</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">extendr_api</a></div><h1>Module <span>conversions</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/extendr_api/conversions/mod.rs.html#1">Source</a> </span></div><h2 id="modules" class="section-header">Modules<a href="#modules" class="anchor">Â§</a></h2><dl class="item-table"><dt><a class="mod" href="try_into_int/index.html" title="mod extendr_api::conversions::try_into_int">try_<wbr>into_<wbr>int</a><span title="Restricted Visibility">&nbsp;ðŸ”’</span> </dt></dl></section></div></main></body></html>