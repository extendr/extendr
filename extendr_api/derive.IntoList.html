<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Derives an implementation of `From&#60;Struct&#62; for Robj` and `From&#60;&#38;Struct&#62; for Robj` on this struct."><title>IntoList in extendr_api - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2"href="../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../static.files/rustdoc-ca0dd0c4.css"><meta name="rustdoc-vars" data-root-path="../" data-static-root-path="../static.files/" data-current-crate="extendr_api" data-themes="" data-resource-suffix="" data-rustdoc-version="1.93.0-nightly (34f954f9b 2025-10-25)" data-channel="nightly" data-search-js="search-d69d8955.js" data-stringdex-js="stringdex-c3e638e9.js" data-settings-js="settings-c38705f0.js" ><script src="../static.files/storage-e2aeef58.js"></script><script defer src="sidebar-items.js"></script><script defer src="../static.files/main-ce535bd0.js"></script><noscript><link rel="stylesheet" href="../static.files/noscript-263c88ec.css"></noscript><link rel="alternate icon" type="image/png" href="../static.files/favicon-32x32-eab170b8.png"><link rel="icon" type="image/svg+xml" href="../static.files/favicon-044be391.svg"></head><body class="rustdoc derive"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><rustdoc-topbar><h2><a href="#">IntoList</a></h2></rustdoc-topbar><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../extendr_api/index.html"><img src="https://raw.githubusercontent.com/extendr/extendr/master/extendr-logo-256.png" alt="logo"></a><h2><a href="../extendr_api/index.html">extendr_<wbr>api</a><span class="version">0.8.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Into<wbr>List</a></h2><h3><a href="#">Sections</a></h3><ul class="block top-toc"><li><a href="#examples" title="Examples">Examples</a></li><li><a href="#details" title="Details">Details</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="index.html">In crate extendr_<wbr>api</a></h2></div></div></nav><div class="sidebar-resizer" title="Drag to resize sidebar"></div><main><div class="width-limiter"><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="index.html">extendr_api</a></div><h1>Derive Macro <span class="derive">Into<wbr>List</span>&nbsp;<button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../src/extendr_macros/lib.rs.html#283">Source</a> </span></div><pre class="rust item-decl"><code>#[derive(IntoList)]
{
    <span class="comment">// Attributes available to this derive:</span>
    #[into_list]
}
</code></pre><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Derives an implementation of <code>From&lt;Struct&gt; for Robj</code> and <code>From&lt;&amp;Struct&gt; for Robj</code> on this struct.</p>
<p>This allows the struct to be converted to a named list in R,
where the list names correspond to the field names of the Rust struct.</p>
<h2 id="examples"><a class="doc-anchor" href="#examples">§</a>Examples</h2>
<p>In the below example, <code>converted</code> contains an R list object with the same fields as the
<code>Foo</code> struct.</p>

<div class="example-wrap ignore"><a href="#" class="tooltip" title="This example is not tested">ⓘ</a><pre class="rust rust-example-rendered"><code><span class="kw">use </span>extendr_api::prelude::<span class="kw-2">*</span>;
<span class="kw">use </span>extendr_macros::IntoList;

<span class="attr">#[derive(IntoList)]
</span><span class="kw">struct </span>Foo {
    a: u32,
    b: String
}
<span class="kw">let </span>converted: Robj = Foo {
    a: <span class="number">5</span>,
    b: String::from(<span class="string">"bar"</span>)
}.into();
<span class="macro">assert_eq!</span>(converted, <span class="macro">R!</span>(<span class="string">r"list(a=5, b='bar')"</span>)<span class="question-mark">?</span>);</code></pre></div>
<p>See <a href="derive.TryFromRobj.html" title="derive extendr_api::TryFromRobj"><code>TryFromRobj</code></a> for a <code>derive</code>-macro in the other direction, i.e.
instantiation of a rust type, by an R list with fields corresponding to
said type.</p>
<p>Supported field attributes</p>
<ul>
<li><code>#[into_list(ignore)]</code> omits the field from being added to the R <code>list()</code></li>
</ul>
<h2 id="details"><a class="doc-anchor" href="#details">§</a>Details</h2>
<p>Note, the <code>From&lt;Struct&gt; for Robj</code> behaviour is different from what is obtained by applying the standard <code>#[extendr]</code> macro
to an <code>impl</code> block. The <code>#[extendr]</code> behaviour returns to R a <strong>pointer</strong> to Rust memory, and generates wrapper functions for calling
Rust functions on that pointer. The implementation from <code>#[derive(IntoList)]</code> actually converts the Rust structure
into a native R list, which allows manipulation and access to internal fields, but it’s a one-way conversion,
and converting it back to Rust will produce a copy of the original struct.</p>
</div></details></section></div></main></body></html>