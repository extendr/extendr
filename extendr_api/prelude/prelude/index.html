<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="ndarray prelude."><title>extendr_api::prelude::prelude - Rust</title><script> if (window.location.protocol !== "file:") document.write(`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceSerif4-Regular-46f98efaafac5295.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Regular-018c141bf0843ffd.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/FiraSans-Medium-8f9a781e4970d388.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Regular-562dcc5011b6de7d.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../../static.files/SourceCodePro-Semibold-d899c5a5c4aeb14a.ttf.woff2">`)</script><link rel="stylesheet" href="../../../static.files/normalize-76eba96aa4d2e634.css"><link rel="stylesheet" href="../../../static.files/rustdoc-e935ef01ae1c1829.css"><meta name="rustdoc-vars" data-root-path="../../../" data-static-root-path="../../../static.files/" data-current-crate="extendr_api" data-themes="" data-resource-suffix="" data-rustdoc-version="1.79.0-nightly (8df7e723e 2024-03-30)" data-channel="nightly" data-search-js="search-42d8da7a6b9792c2.js" data-settings-js="settings-4313503d2e1961c2.js" ><script src="../../../static.files/storage-4c98445ec4002617.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../../static.files/main-12cf3b4f4f9dc36d.js"></script><noscript><link rel="stylesheet" href="../../../static.files/noscript-04d5337699b92874.css"></noscript><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-16x16-8b506e7a72182f1c.png"><link rel="alternate icon" type="image/png" href="../../../static.files/favicon-32x32-422f7d1d52889060.png"><link rel="icon" type="image/svg+xml" href="../../../static.files/favicon-2c020d218678b618.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button><a class="logo-container" href="../../../extendr_api/index.html"><img src="https://raw.githubusercontent.com/extendr/extendr/master/extendr-logo-256.png" alt=""></a></nav><nav class="sidebar"><div class="sidebar-crate"><a class="logo-container" href="../../../extendr_api/index.html"><img src="https://raw.githubusercontent.com/extendr/extendr/master/extendr-logo-256.png" alt="logo"></a><h2><a href="../../../extendr_api/index.html">extendr_api</a><span class="version">0.6.0</span></h2></div><h2 class="location"><a href="#">Module prelude</a></h2><div class="sidebar-elems"><section><ul class="block"><li><a href="#macros">Macros</a></li><li><a href="#structs">Structs</a></li><li><a href="#traits">Traits</a></li><li><a href="#functions">Functions</a></li><li><a href="#types">Type Aliases</a></li></ul></section><h2><a href="../index.html">In extendr_api::prelude</a></h2></div></nav><div class="sidebar-resizer"></div>
    <main><div class="width-limiter"><nav class="sub"><form class="search-form"><span></span><div id="sidebar-button" tabindex="-1"><a href="../../../extendr_api/all.html" title="show sidebar"></a></div><input class="search-input" name="search" aria-label="Run search in the documentation" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" tabindex="-1"><a href="../../../help.html" title="help">?</a></div><div id="settings-menu" tabindex="-1"><a href="../../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../../static.files/wheel-7b819b6101059cd0.svg"></a></div></form></nav><section id="main-content" class="content"><div class="main-heading"><h1>Module <a href="../../index.html">extendr_api</a>::<wbr><a href="../index.html">prelude</a>::<wbr><a class="mod" href="#">prelude</a><button id="copy-path" title="Copy item path to clipboard"><img src="../../../static.files/clipboard-7571035ce49a181d.svg" width="19" height="18" alt="Copy item path"></button></h1><span class="out-of-band"><a class="src" href="https://docs.rs/ndarray/0.15/src/ndarray/lib.rs.html#225">source</a> · <button id="toggle-all-docs" title="collapse all docs">[<span>&#x2212;</span>]</button></span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>ndarray prelude.</p>
<p>This module contains the most used types, type aliases, traits, functions,
and macros that you can import easily as a group.</p>

<div class="example-wrap"><pre class="rust rust-example-rendered"><code><span class="kw">use </span>ndarray::prelude::<span class="kw-2">*</span>;
</code></pre></div>
</div></details><h2 id="macros" class="section-header">Macros<a href="#macros" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="macro" href="macro.array.html" title="macro extendr_api::prelude::prelude::array">array</a></div><div class="desc docblock-short">Create an <strong><a href="../type.Array.html" title="type extendr_api::prelude::Array"><code>Array</code></a></strong> with one, two or
three dimensions.</div></li><li><div class="item-name"><a class="macro" href="macro.azip.html" title="macro extendr_api::prelude::prelude::azip">azip</a></div><div class="desc docblock-short">Array zip macro: lock step function application across several arrays and
producers.</div></li><li><div class="item-name"><a class="macro" href="macro.s.html" title="macro extendr_api::prelude::prelude::s">s</a></div><div class="desc docblock-short">Slice argument constructor.</div></li></ul><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="struct" href="struct.ArrayBase.html" title="struct extendr_api::prelude::prelude::ArrayBase">ArrayBase</a></div><div class="desc docblock-short">An <em>n</em>-dimensional array.</div></li><li><div class="item-name"><a class="struct" href="struct.Axis.html" title="struct extendr_api::prelude::prelude::Axis">Axis</a></div><div class="desc docblock-short">An axis index.</div></li><li><div class="item-name"><a class="struct" href="struct.Dim.html" title="struct extendr_api::prelude::prelude::Dim">Dim</a></div><div class="desc docblock-short">Dimension description.</div></li><li><div class="item-name"><a class="struct" href="struct.NewAxis.html" title="struct extendr_api::prelude::prelude::NewAxis">NewAxis</a></div><div class="desc docblock-short">Token to represent a new axis in a slice description.</div></li></ul><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="trait" href="trait.AsArray.html" title="trait extendr_api::prelude::prelude::AsArray">AsArray</a></div><div class="desc docblock-short">Argument conversion into an array view</div></li><li><div class="item-name"><a class="trait" href="trait.Dimension.html" title="trait extendr_api::prelude::prelude::Dimension">Dimension</a></div><div class="desc docblock-short">Array shape and index trait.</div></li><li><div class="item-name"><a class="trait" href="trait.NdFloat.html" title="trait extendr_api::prelude::prelude::NdFloat">NdFloat</a></div><div class="desc docblock-short">Floating-point element types <code>f32</code> and <code>f64</code>.</div></li><li><div class="item-name"><a class="trait" href="trait.ShapeBuilder.html" title="trait extendr_api::prelude::prelude::ShapeBuilder">ShapeBuilder</a></div><div class="desc docblock-short">A trait for <code>Shape</code> and <code>D where D: Dimension</code> that allows
customizing the memory layout (strides) of an array shape.</div></li></ul><h2 id="functions" class="section-header">Functions<a href="#functions" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="fn" href="fn.Dim.html" title="fn extendr_api::prelude::prelude::Dim">Dim</a></div><div class="desc docblock-short">Create a new dimension value.</div></li><li><div class="item-name"><a class="fn" href="fn.Ix0.html" title="fn extendr_api::prelude::prelude::Ix0">Ix0</a></div><div class="desc docblock-short">Create a zero-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.Ix1.html" title="fn extendr_api::prelude::prelude::Ix1">Ix1</a></div><div class="desc docblock-short">Create a one-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.Ix2.html" title="fn extendr_api::prelude::prelude::Ix2">Ix2</a></div><div class="desc docblock-short">Create a two-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.Ix3.html" title="fn extendr_api::prelude::prelude::Ix3">Ix3</a></div><div class="desc docblock-short">Create a three-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.Ix4.html" title="fn extendr_api::prelude::prelude::Ix4">Ix4</a></div><div class="desc docblock-short">Create a four-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.Ix5.html" title="fn extendr_api::prelude::prelude::Ix5">Ix5</a></div><div class="desc docblock-short">Create a five-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.Ix6.html" title="fn extendr_api::prelude::prelude::Ix6">Ix6</a></div><div class="desc docblock-short">Create a six-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.IxDyn.html" title="fn extendr_api::prelude::prelude::IxDyn">IxDyn</a></div><div class="desc docblock-short">Create a dynamic-dimensional index</div></li><li><div class="item-name"><a class="fn" href="fn.arr0.html" title="fn extendr_api::prelude::prelude::arr0">arr0</a></div><div class="desc docblock-short">Create a zero-dimensional array with the element <code>x</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.arr1.html" title="fn extendr_api::prelude::prelude::arr1">arr1</a></div><div class="desc docblock-short">Create a one-dimensional array with elements from <code>xs</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.arr2.html" title="fn extendr_api::prelude::prelude::arr2">arr2</a></div><div class="desc docblock-short">Create a two-dimensional array with elements from <code>xs</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.aview0.html" title="fn extendr_api::prelude::prelude::aview0">aview0</a></div><div class="desc docblock-short">Create a zero-dimensional array view borrowing <code>x</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.aview1.html" title="fn extendr_api::prelude::prelude::aview1">aview1</a></div><div class="desc docblock-short">Create a one-dimensional array view with elements borrowing <code>xs</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.aview2.html" title="fn extendr_api::prelude::prelude::aview2">aview2</a></div><div class="desc docblock-short">Create a two-dimensional array view with elements borrowing <code>xs</code>.</div></li><li><div class="item-name"><a class="fn" href="fn.aview_mut1.html" title="fn extendr_api::prelude::prelude::aview_mut1">aview_mut1</a></div><div class="desc docblock-short">Create a one-dimensional read-write array view with elements borrowing <code>xs</code>.</div></li></ul><h2 id="types" class="section-header">Type Aliases<a href="#types" class="anchor">§</a></h2><ul class="item-table"><li><div class="item-name"><a class="type" href="type.ArcArray.html" title="type extendr_api::prelude::prelude::ArcArray">ArcArray</a></div><div class="desc docblock-short">An array where the data has shared ownership and is copy on write.</div></li><li><div class="item-name"><a class="type" href="type.Array.html" title="type extendr_api::prelude::prelude::Array">Array</a></div><div class="desc docblock-short">An array that owns its data uniquely.</div></li><li><div class="item-name"><a class="type" href="type.Array0.html" title="type extendr_api::prelude::prelude::Array0">Array0</a></div><div class="desc docblock-short">zero-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.Array1.html" title="type extendr_api::prelude::prelude::Array1">Array1</a></div><div class="desc docblock-short">one-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.Array2.html" title="type extendr_api::prelude::prelude::Array2">Array2</a></div><div class="desc docblock-short">two-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.Array3.html" title="type extendr_api::prelude::prelude::Array3">Array3</a></div><div class="desc docblock-short">three-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.Array4.html" title="type extendr_api::prelude::prelude::Array4">Array4</a></div><div class="desc docblock-short">four-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.Array5.html" title="type extendr_api::prelude::prelude::Array5">Array5</a></div><div class="desc docblock-short">five-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.Array6.html" title="type extendr_api::prelude::prelude::Array6">Array6</a></div><div class="desc docblock-short">six-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.ArrayD.html" title="type extendr_api::prelude::prelude::ArrayD">ArrayD</a></div><div class="desc docblock-short">dynamic-dimensional array</div></li><li><div class="item-name"><a class="type" href="type.ArrayView.html" title="type extendr_api::prelude::prelude::ArrayView">ArrayView</a></div><div class="desc docblock-short">A read-only array view.</div></li><li><div class="item-name"><a class="type" href="type.ArrayView0.html" title="type extendr_api::prelude::prelude::ArrayView0">ArrayView0</a></div><div class="desc docblock-short">zero-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayView1.html" title="type extendr_api::prelude::prelude::ArrayView1">ArrayView1</a></div><div class="desc docblock-short">one-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayView2.html" title="type extendr_api::prelude::prelude::ArrayView2">ArrayView2</a></div><div class="desc docblock-short">two-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayView3.html" title="type extendr_api::prelude::prelude::ArrayView3">ArrayView3</a></div><div class="desc docblock-short">three-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayView4.html" title="type extendr_api::prelude::prelude::ArrayView4">ArrayView4</a></div><div class="desc docblock-short">four-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayView5.html" title="type extendr_api::prelude::prelude::ArrayView5">ArrayView5</a></div><div class="desc docblock-short">five-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayView6.html" title="type extendr_api::prelude::prelude::ArrayView6">ArrayView6</a></div><div class="desc docblock-short">six-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewD.html" title="type extendr_api::prelude::prelude::ArrayViewD">ArrayViewD</a></div><div class="desc docblock-short">dynamic-dimensional array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut.html" title="type extendr_api::prelude::prelude::ArrayViewMut">ArrayViewMut</a></div><div class="desc docblock-short">A read-write array view.</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut0.html" title="type extendr_api::prelude::prelude::ArrayViewMut0">ArrayViewMut0</a></div><div class="desc docblock-short">zero-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut1.html" title="type extendr_api::prelude::prelude::ArrayViewMut1">ArrayViewMut1</a></div><div class="desc docblock-short">one-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut2.html" title="type extendr_api::prelude::prelude::ArrayViewMut2">ArrayViewMut2</a></div><div class="desc docblock-short">two-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut3.html" title="type extendr_api::prelude::prelude::ArrayViewMut3">ArrayViewMut3</a></div><div class="desc docblock-short">three-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut4.html" title="type extendr_api::prelude::prelude::ArrayViewMut4">ArrayViewMut4</a></div><div class="desc docblock-short">four-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut5.html" title="type extendr_api::prelude::prelude::ArrayViewMut5">ArrayViewMut5</a></div><div class="desc docblock-short">five-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMut6.html" title="type extendr_api::prelude::prelude::ArrayViewMut6">ArrayViewMut6</a></div><div class="desc docblock-short">six-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.ArrayViewMutD.html" title="type extendr_api::prelude::prelude::ArrayViewMutD">ArrayViewMutD</a></div><div class="desc docblock-short">dynamic-dimensional read-write array view</div></li><li><div class="item-name"><a class="type" href="type.CowArray.html" title="type extendr_api::prelude::prelude::CowArray">CowArray</a></div><div class="desc docblock-short">An array with copy-on-write behavior.</div></li><li><div class="item-name"><a class="type" href="type.Ix0.html" title="type extendr_api::prelude::prelude::Ix0">Ix0</a></div><div class="desc docblock-short">zero-dimensionial</div></li><li><div class="item-name"><a class="type" href="type.Ix1.html" title="type extendr_api::prelude::prelude::Ix1">Ix1</a></div><div class="desc docblock-short">one-dimensional</div></li><li><div class="item-name"><a class="type" href="type.Ix2.html" title="type extendr_api::prelude::prelude::Ix2">Ix2</a></div><div class="desc docblock-short">two-dimensional</div></li><li><div class="item-name"><a class="type" href="type.Ix3.html" title="type extendr_api::prelude::prelude::Ix3">Ix3</a></div><div class="desc docblock-short">three-dimensional</div></li><li><div class="item-name"><a class="type" href="type.Ix4.html" title="type extendr_api::prelude::prelude::Ix4">Ix4</a></div><div class="desc docblock-short">four-dimensional</div></li><li><div class="item-name"><a class="type" href="type.Ix5.html" title="type extendr_api::prelude::prelude::Ix5">Ix5</a></div><div class="desc docblock-short">five-dimensional</div></li><li><div class="item-name"><a class="type" href="type.Ix6.html" title="type extendr_api::prelude::prelude::Ix6">Ix6</a></div><div class="desc docblock-short">six-dimensional</div></li><li><div class="item-name"><a class="type" href="type.IxDyn.html" title="type extendr_api::prelude::prelude::IxDyn">IxDyn</a></div><div class="desc docblock-short">dynamic-dimensional</div></li><li><div class="item-name"><a class="type" href="type.RawArrayView.html" title="type extendr_api::prelude::prelude::RawArrayView">RawArrayView</a></div><div class="desc docblock-short">A read-only array view without a lifetime.</div></li><li><div class="item-name"><a class="type" href="type.RawArrayViewMut.html" title="type extendr_api::prelude::prelude::RawArrayViewMut">RawArrayViewMut</a></div><div class="desc docblock-short">A mutable array view without a lifetime.</div></li></ul></section></div></main></body></html>