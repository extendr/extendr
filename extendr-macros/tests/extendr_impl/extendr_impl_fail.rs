#![allow(unused_imports)]

#[allow(unused_imports)]
use extendr_api::prelude::*;

struct Foo;

#[extendr]
impl Foo {
    fn take_ownership(self) {}
}

struct Foo2;

#[extendr]
impl Foo2 {
    fn take_ownership(self: Self) {}
}

struct Foo3;

#[extendr]
impl Foo3 {
    fn take_ownership(self: Foo3) {}
}
struct Foo4;

#[extendr]
impl Foo4 {
    fn return_ownership(self) -> Self {}
}

struct Foo5;

#[extendr]
impl Foo5 {
    fn take_ownership(self: Self) -> Foo5 {}
}

fn main() {}
