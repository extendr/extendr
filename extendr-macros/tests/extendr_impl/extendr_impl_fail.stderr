error: expected &self or &mut self
  --> tests/extendr_impl/extendr_impl_fail.rs:10:23
   |
10 |     fn take_ownership(self) {}
   |                       ^^^^

error: expected &self or &mut self
  --> tests/extendr_impl/extendr_impl_fail.rs:17:23
   |
17 |     fn take_ownership(self: Self) {}
   |                       ^^^^^^^^^^

error: expected &self or &mut self
  --> tests/extendr_impl/extendr_impl_fail.rs:24:23
   |
24 |     fn take_ownership(self: Foo3) {}
   |                       ^^^^^^^^^^

error: expected &self or &mut self
  --> tests/extendr_impl/extendr_impl_fail.rs:30:25
   |
30 |     fn return_ownership(self) -> Self {}
   |                         ^^^^

error: expected &self or &mut self
  --> tests/extendr_impl/extendr_impl_fail.rs:37:23
   |
37 |     fn take_ownership(self: Self) -> Foo5 {}
   |                       ^^^^^^^^^^
